lib/phoenix_live_view/utils.ex

```diff
@@ -93,19 +93,77 @@ defmodule Phoenix.LiveView.Utils do
   @doc """
   Forces an assign with the given changed map.
   """
-  def force_assign(assigns, nil, key, val), do: Map.put(assigns, key, val)
+  def force_assign(assigns, nil, key, val),
+    do: Map.put(assigns, key, maybe_check_expensive_assign(key, val))
 
   def force_assign(assigns, changed, key, val) do
     # If the current value is a map, we store it in changed so
     # we can perform nested change tracking. Also note the use
     # of put_new is important. We want to keep the original value
     # from assigns and not any intermediate ones that may appear.
     current_val = Map.get(assigns, key)
+    maybe_check_expensive_assign(key, val)
     changed_val = if is_map(current_val), do: current_val, else: true
     changed = Map.put_new(changed, key, changed_val)
     Map.put(%{assigns | __changed__: changed}, key, val)
   end
 
+  if Application.compile_env(:phoenix_live_view, :warn_on_expensive_assigns, false) do
+    defp maybe_check_expensive_assign(key, %Phoenix.LiveView.Socket{}) do
+      IO.warn("""
+      you are assigning the LiveView socket itself into the assign #{key}.
+
+      Because of LiveView's change tracking this can lead to more memory usage than expected,
+      as old data won't be cleared in the assign.
+      """)
+    end
+
+    defp maybe_check_expensive_assign(key, %{__changed__: _}) do
+      IO.warn("""
+      you are assigning an assigns map into the assign #{key}.
+
+      Because of LiveView's change tracking this can lead to more memory usage than expected,
+      as old data won't be cleared in the assign.
+      """)
+    end
+
+    defp maybe_check_expensive_assign(_key, val) when is_map(val) and not is_struct(val) do
+      Enum.each(val, fn {key, val} ->
+        maybe_check_expensive_assign(key, val)
```

Comment:
Maybe passing the original key would be better for the message. Or building a path. Not sure yet.